[xdp-tools] add initial specfile at 1.2.9

dankamongmen opened the pull request:
replaces #4512, of which we shall not speak further. this is against `main`, whereas that was against `2.0`.

Taken from xdp-tools' packaging/rpm/xdp-tools.spec and made to work with Mariner. This required disabling binary stripping and replacing 'emacs-nox' with 'emacs' in BuildRequires.

###### Merge Checklist  <!-- REQUIRED -->
<!-- You can set them now ([x]) or set them later using the Github UI -->
**All** boxes should be checked before merging the PR *(just tick any boxes which don't apply to this PR)*
- [X] The toolchain has been rebuilt successfully (or no changes were made to it)
- [X] The toolchain/worker package manifests are up-to-date
- [x] Any updated packages successfully build (or no packages were changed)
- [x] Packages depending on static components modified in this PR (Golang, `*-static` subpackages, etc.) have had their `Release` tag incremented.
- [x] Package tests (%check section) have been verified with RUN_CHECK=y for existing SPEC files, or added to new SPEC files
- [x] All package sources are available
- [x] cgmanifest files are up-to-date and sorted (`./cgmanifest.json`, `./toolkit/scripts/toolchain/cgmanifest.json`, `.github/workflows/cgmanifest.json`)
- [x] LICENSE-MAP files are up-to-date (`./SPECS/LICENSES-AND-NOTICES/data/licenses.json`, `./SPECS/LICENSES-AND-NOTICES/LICENSES-MAP.md`, `./SPECS/LICENSES-AND-NOTICES/LICENSE-EXCEPTIONS.PHOTON`)
- [x] All source files have up-to-date hashes in the `*.signatures.json` files
- [x] `sudo make go-tidy-all` and `sudo make go-test-coverage` pass
- [x] Documentation has been updated to match any changes to the build system
- [x] Ready to merge

---

###### Summary 
Adds `xdp-tools` specfile to Mariner 2.0 (#4408). This is necessary to use libxdp and related tools.

###### Change Log  
Took the Fedora `spec`file and made two changes to build on Mariner:
 * disabled stripping (with FIXME comment)
 * replaced `emacs-nox` BuildRequires with `emacs`

###### Does this affect the toolchain?  
No

###### Associated issues
fixes #4408 

###### Links to CVEs 

###### Test Methodology
Local build in Mariner 2.0 VM

mfrw commented on SPECS/xdp-tools/xdp-tools.spec:
```diff
@@ -0,0 +1,157 @@
+Summary:        Utilities and example programs for use with XDP
+Name:           xdp-tools
```
Thank you for your contribution.

Initial comments:

- We are missing a mandatory `xdp-tools.signature.json` file; which, if absent will break the build :(
- We need to upload the source `tar.gz` to the `blobstore` 

I have un-checked above items from the checklist in the PR.

Please feel free to reach-out in case you need any help :)

dankamongmen replied:
so i need to move this from SPECS to SPECS-EXTENDED and generate the signature.json, right?

dankamongmen replied:
what is this blobstore? do i have access to it?

dankamongmen replied:
i've moved the specfile over to `SPECS-EXTENDED` and added a .signature.json. still unsure about this blobstore.

oliviacrain replied:
I went ahead and uploaded the source tarball to our blobstore. Quick explanation- our SRPM packing tool doesn't pull sources directly from the source URL. Instead, we archive all sources and make them available in a known location (`cblmarinerstorage.blob.core.windows.net/sources/core`). That way, we don't ever have to deal with build breaks from source URLs going offline.

oliviacrain commented on SPECS-EXTENDED/xdp-tools/xdp-tools.spec:
```diff
@@ -0,0 +1,157 @@
+Name:           xdp-tools
+Version:        1.2.9
+Release:        1%{?dist}
+License:        GPLv2, LGPLv2, BSD
```
We're moving towards using [SPDX expressions](https://spdx.org/licenses/) for our license tags. I went ahead and double-checked that the licenses present in this repo correspond to the following expression:

```suggestion
License:        GPL-2.0-only AND LGPL-2.1-only AND BSD-2-Clause
```

dankamongmen replied:
done, thanks!

oliviacrain commented:
Looks great overall- going to kick off a couple of buddy builds to validate, should have results in a few minutes.

dankamongmen commented:
thanks! yeah, i gathered the purpose of the blobstore =]. my question was more one of "do i have access to this storehouse of blobbery"

oliviacrain commented:
Ah, yeah, nope! I don't think we give publish access to anyone outside our core teams.

oliviacrain commented:
Sorry for the delay- it's been a busy week for our team. Will re-run validation now and get the results back soon.

oliviacrain commented on SPECS-EXTENDED/xdp-tools/xdp-tools.signatures.json:
```diff
@@ -0,0 +1,5 @@
+{
+ "Signatures": {
+  "xdp-tools-1.2.9.tar.gz": "3f8d30bd2e00c522103d224014f59a95400144aba1f3b322c6ad473541a0f99e",
```
Looks like the SRPM packer doesn't like the trailing comma here. Logs are below:

```
ERRO[0006] Failed to read the signatures file (/specs/xdp-tools/xdp-tools.signatures.json): invalid character '}' looking for beginning of object key string. 
ERRO[0006] Failed to pack (/specs/xdp-tools/xdp-tools.spec). Error: invalid character '}' looking for beginning of object key string 
PANI[0006] invalid character '}' looking for beginning of object key string 

```

oliviacrain commented:
Hm, looks like the symbol stripping doesn't like some of the object files produced by this package. See the link below for the build log.

https://gist.github.com/oliviacrain/04a1b3e4ff6f838ee6cd60aac3bfc656

dankamongmen commented:
> Hm, looks like the symbol stripping doesn't like some of the object files produced by this package. See the link below for the build log.
> 
> https://gist.github.com/oliviacrain/04a1b3e4ff6f838ee6cd60aac3bfc656

thanks, taking a look now. i ought be able to bring it into conformance if that output is generated by the autotests linked here. let me also resolve this merge conflict.

dankamongmen commented:
> Hm, looks like the symbol stripping doesn't like some of the object files produced by this package. See the link below for the build log.
> 
> https://gist.github.com/oliviacrain/04a1b3e4ff6f838ee6cd60aac3bfc656

ahhh yes, i have this addressed in our local copy; not sure how it got lost on the way here.

fixed. let's see if the linter has feedback. PTAL, thanks!

dankamongmen commented:
> fixed. let's see if the linter has feedback. PTAL, thanks!

ahhh, nice catch from the linter. alright, this ought be good now.

dankamongmen commented:
alright, all checks now, and your build problem ought be resolved--sorry about that.

we can update this whenever mariner releases a full eBPF-capable toolchain (i suspect the problem to be related to BTF, but am unsure). for now, it's a small wart.

it says there's an unresolved comment, but i don't see any? i might be stupid.

anyway, PTAL, and thanks for all your help and thoughtful review!

oliviacrain commented:
Taking a look and kicked off another validation build. Will report back shortly.

oliviacrain commented on SPECS-EXTENDED/xdp-tools/xdp-tools.spec:
```diff
@@ -0,0 +1,161 @@
+# find-debuginfo produces empty debugsourcefiles.list
+# disable the debug package to avoid rpmbuild error'ing out because of this
+%global debug_package %{nil}
+%global __os_install_post %{_lib}/rpm/brp-compress %{nil}
```
Still getting a build failure, due to `%{_lib}/rpm/brp-compress` expanding to `lib/rpm/brp-compress`, which doesn't exist in the context of the working directory here.

This doesn't quite work with [the way we structure our post-install scripts](https://github.com/microsoft/CBL-Mariner/blob/main/SPECS/mariner-rpm-macros/macros#L72-L87)- changing `%__os_install_post` like this is taking an axe to something that just needs a chisel. Defining `%{__strip}` in the following way should fix the issue:

```suggestion
%global __strip /bin/true
```

dankamongmen replied:
cheerfully applied, thanks! odd, i wonder why this is working in our mariner build.

oliviacrain commented:
Good news- post-install phase passes now. Bad news- `%files` section fails for `xdp-tools`. Not sure why your build environment is passing where ours is being more strict. Full log is in in [this gist](https://gist.github.com/oliviacrain/7280b02952761d608db1a514585d6049).

dankamongmen commented:
> Good news- post-install phase passes now. Bad news- `%files` section fails. Not sure why your build environment is passing where ours is being more strict. Full log is in in [this gist](https://gist.github.com/oliviacrain/7280b02952761d608db1a514585d6049).

it's another place where my two specfiles had diverged; sorry for wasting your time. i've removed these entries, and also added some new ones that would otherwise have led to a subsequent failure.

https://msazure.visualstudio.com/One/_git/Azure-Orbital-Garuda?path=/src/xdp-tools.spec

dankamongmen commented:
synced to upstream/main, eliminating merge conflicts

dankamongmen commented:
> Good news- post-install phase passes now. Bad news- `%files` section fails. Not sure why your build environment is passing where ours is being more strict. Full log is in in [this gist](https://gist.github.com/oliviacrain/7280b02952761d608db1a514585d6049).

am i correct in reading the autochecks to now perform these basic tests? hurrah, nice work Mariner team!

dankamongmen commented:
imma go ahead and rebase this branch and pull in the newest version (1.3.1)
